plugin.rb

```diff
@@ -146,14 +146,21 @@ def index
     end
   end
 
+  add_to_class(:user, :can_use_canned_replies?) do
+    return true if staff?
+    group_list = SiteSetting.canned_replies_groups.split("|").map!(&:downcase)
+    groups.any? { |group| group_list.include?(group.name.downcase) }
```

Comment:
I've always wondered if doing this like this is any faster/slower?

```ruby
(groups.map { |g| g.name.downcase } & group_list).present?
```

Reply:
My way looks a lot faster.

```ruby
# frozen_string_literal: true

require 'benchmark/ips'
require File.expand_path("../../config/environment", __FILE__)

groups = [
  OpenStruct.new(name: "staff"),
  OpenStruct.new(name: "admins"),
  OpenStruct.new(name: "moderators"),
  OpenStruct.new(name: "trust_level_0"),
  OpenStruct.new(name: "trust_level_1"),
  OpenStruct.new(name: "trust_level_2"),
  OpenStruct.new(name: "trust_level_3"),
  OpenStruct.new(name: "secret"),
]

group_list = ["secret", "staff"]

Benchmark.ips do |b|
  b.report("using any?") do
    groups.any? { |group| group_list.include?(group.name.downcase) }
  end

  b.report("using &.present?") do
    (groups.map { |g| g.name.downcase } & group_list).present?
  end
end
```

produces

```
Warming up --------------------------------------
          using any?   208.773k i/100ms
    using &.present?    41.144k i/100ms
Calculating -------------------------------------
          using any?      3.687M (± 1.4%) i/s -     18.581M in   5.040291s
    using &.present?    465.854k (± 1.9%) i/s -      2.345M in   5.035978s
```

Reply:
That makes sense since `.any?` will stop as soon as possible.